File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_privacy.php
<?php
declare(strict_types=1);

// Routes: campaign-scoped privacy controls -- role-filtered notes,
// character-to-character whispers, and basic character sheets.
// ---------------------------------------------------------------------------

function play_campaign_is_member(PDO $db, string $campaignId, string $username): bool {
    $stmt = $db->prepare('SELECT 1 FROM play_campaign_members WHERE campaign_id = ? AND username = ?');
    $stmt->execute([$campaignId, $username]);
    return $stmt->fetch(PDO::FETCH_ASSOC) !== false;
}

// Enforces the baseline visibility rule shared by every privacy-controls
// route: the actor must be the campaign DM or a campaign member. Returns
// whether the actor is the DM; sends 403 and exits otherwise.
function require_campaign_access(PDO $db, array $campaign, array $actor): bool {
    $isDm = $actor['username'] === $campaign['owner'];
    if (!$isDm && !play_campaign_is_member($db, $campaign['id'], $actor['username'])) {
        forbidden('only the dm or a campaign member may access this campaign');
    }
    return $isDm;
}

function next_note_sequence(PDO $db, string $campaignId): int {
    $stmt = $db->prepare('SELECT COALESCE(MAX(seq), 0) AS max_seq FROM play_campaign_notes WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    return (int)$stmt->fetch(PDO::FETCH_ASSOC)['max_seq'] + 1;
}

function next_whisper_sequence(PDO $db, string $campaignId): int {
    $stmt = $db->prepare('SELECT COALESCE(MAX(seq), 0) AS max_seq FROM play_campaign_whispers WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    return (int)$stmt->fetch(PDO::FETCH_ASSOC)['max_seq'] + 1;
}

function is_valid_visibility($value): bool {
    return is_string($value) && in_array($value, ['private', 'party'], true);
}

// ---------------------------------------------------------------------------
// Notes
// ---------------------------------------------------------------------------

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/notes$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if (!play_campaign_is_member($db, $campaignId, $actor['username'])) {
        forbidden('only a campaign member may create notes');
    }

    $body = read_json_body();
    if ($body === null || !isset($body['note_id'], $body['text'], $body['visibility'])
        || !is_string($body['note_id']) || $body['note_id'] === ''
        || !is_string($body['text']) || $body['text'] === ''
        || !is_valid_visibility($body['visibility'])) {
        bad_request();
    }
    $noteId = $body['note_id'];

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_notes WHERE campaign_id = ? AND note_id = ?');
    $stmt->execute([$campaignId, $noteId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('note id already exists');
    }

    $note = [
        'note_id' => $noteId,
        'text' => $body['text'],
        'visibility' => $body['visibility'],
        'owner' => $actor['username'],
    ];

    $seq = next_note_sequence($db, $campaignId);
    $stmt = $db->prepare('INSERT INTO play_campaign_notes (campaign_id, note_id, seq, data) VALUES (?, ?, ?, ?)');
    $stmt->execute([$campaignId, $noteId, $seq, json_encode($note)]);

    send_json($note, 201);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/notes$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);
    $isDm = require_campaign_access($db, $campaign, $actor);

    $stmt = $db->prepare('SELECT data FROM play_campaign_notes WHERE campaign_id = ? ORDER BY seq ASC');
    $stmt->execute([$campaignId]);
    $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);

    $notes = array_map(function ($r) {
        return json_decode($r['data'], true);
    }, $rows);

    if (!$isDm) {
        $notes = array_values(array_filter($notes, function ($note) use ($actor) {
            if ($note['visibility'] === 'party') {
                return true;
            }
            return $note['owner'] === $actor['username'];
        }));
    }

    send_json(['notes' => array_values($notes)]);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/notes/([^/]+)$#', $path, $pm)) {
    $campaignId = $pm[1];
    $noteId = $pm[2];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);
    $isDm = require_campaign_access($db, $campaign, $actor);

    $stmt = $db->prepare('SELECT data FROM play_campaign_notes WHERE campaign_id = ? AND note_id = ?');
    $stmt->execute([$campaignId, $noteId]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row === false) {
        not_found('note not found');
    }
    $note = json_decode($row['data'], true);

    if (!$isDm && $note['visibility'] === 'private' && $note['owner'] !== $actor['username']) {
        forbidden('this note is private to its owner');
    }

    send_json($note);
}

if ($method === 'PUT' && preg_match('#^/v1/play/campaigns/([^/]+)/notes/([^/]+)$#', $path, $pm)) {
    $campaignId = $pm[1];
    $noteId = $pm[2];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);
    require_campaign_access($db, $campaign, $actor);

    $stmt = $db->prepare('SELECT data FROM play_campaign_notes WHERE campaign_id = ? AND note_id = ?');
    $stmt->execute([$campaignId, $noteId]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row === false) {
        not_found('note not found');
    }
    $note = json_decode($row['data'], true);

    if ($note['owner'] !== $actor['username']) {
        forbidden('only the note owner may update this note');
    }

    $body = read_json_body();
    if ($body === null || !isset($body['text'], $body['visibility'])
        || !is_string($body['text']) || $body['text'] === ''
        || !is_valid_visibility($body['visibility'])) {
        bad_request();
    }

    $note['text'] = $body['text'];
    $note['visibility'] = $body['visibility'];

    $stmt = $db->prepare('UPDATE play_campaign_notes SET data = ? WHERE campaign_id = ? AND note_id = ?');
    $stmt->execute([json_encode($note), $campaignId, $noteId]);

    send_json($note);
}

// ---------------------------------------------------------------------------
// Whispers
// ---------------------------------------------------------------------------

// Finds the character_id currently owned by a username, or null if the
// actor is not a player with an owned character in this campaign.
function play_actor_owned_character(PDO $db, string $campaignId, string $username): ?string {
    $stmt = $db->prepare('SELECT username, character_id FROM play_campaign_members WHERE campaign_id = ? ORDER BY rowid ASC');
    $stmt->execute([$campaignId]);
    $members = $stmt->fetchAll(PDO::FETCH_ASSOC);
    foreach ($members as $member) {
        $owner = play_character_owner($db, $campaignId, $member['character_id'], $member['username']);
        if ($owner === $username) {
            return $member['character_id'];
        }
    }
    return null;
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/whispers$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if (!play_campaign_is_member($db, $campaignId, $actor['username'])) {
        forbidden('only a campaign player may send whispers');
    }

    $fromCharacterId = play_actor_owned_character($db, $campaignId, $actor['username']);
    if ($fromCharacterId === null) {
        forbidden('only a campaign player with an owned character may send whispers');
    }

    $body = read_json_body();
    if ($body === null || !isset($body['whisper_id'], $body['to_character_id'], $body['text'])
        || !is_string($body['whisper_id']) || $body['whisper_id'] === ''
        || !is_string($body['to_character_id']) || $body['to_character_id'] === ''
        || !is_string($body['text']) || $body['text'] === '') {
        bad_request();
    }
    $whisperId = $body['whisper_id'];
    $toCharacterId = $body['to_character_id'];

    if (find_play_character($db, $campaignId, $toCharacterId) === null) {
        bad_request('to_character_id must belong to a current campaign member');
    }

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_whispers WHERE campaign_id = ? AND whisper_id = ?');
    $stmt->execute([$campaignId, $whisperId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('whisper id already exists');
    }

    $whisper = [
        'whisper_id' => $whisperId,
        'from_character_id' => $fromCharacterId,
        'to_character_id' => $toCharacterId,
        'text' => $body['text'],
    ];

    $seq = next_whisper_sequence($db, $campaignId);
    $stmt = $db->prepare('INSERT INTO play_campaign_whispers (campaign_id, whisper_id, seq, data) VALUES (?, ?, ?, ?)');
    $stmt->execute([$campaignId, $whisperId, $seq, json_encode($whisper)]);

    send_json($whisper, 201);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/whispers$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);
    $isDm = require_campaign_access($db, $campaign, $actor);

    $stmt = $db->prepare('SELECT data FROM play_campaign_whispers WHERE campaign_id = ? ORDER BY seq ASC');
    $stmt->execute([$campaignId]);
    $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);

    $whispers = array_map(function ($r) {
        return json_decode($r['data'], true);
    }, $rows);

    if (!$isDm) {
        $ownedCharacterId = play_actor_owned_character($db, $campaignId, $actor['username']);
        $whispers = array_values(array_filter($whispers, function ($whisper) use ($ownedCharacterId) {
            return $ownedCharacterId !== null
                && ($whisper['from_character_id'] === $ownedCharacterId || $whisper['to_character_id'] === $ownedCharacterId);
        }));
    }

    send_json(['whispers' => array_values($whispers)]);
}

// ---------------------------------------------------------------------------
// Character sheets
// ---------------------------------------------------------------------------

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/characters/([^/]+)/sheet$#', $path, $pm)) {
    $campaignId = $pm[1];
    $charId = $pm[2];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $found = find_play_character($db, $campaignId, $charId);
    if ($found === null) {
        not_found('character not found');
    }
    [$creatorUsername, $member] = $found;

    $isDm = $actor['username'] === $campaign['owner'];
    $owner = play_character_owner($db, $campaignId, $charId, $creatorUsername);

    if (!$isDm && $actor['username'] !== $owner) {
        forbidden('only the character owner or the dm may view this sheet');
    }

    send_json([
        'character_id' => $charId,
        'owner' => $owner,
        'name' => $member['name'],
        'class' => $member['class'],
        'level' => 1,
        'proficiency_bonus' => 2,
        'hp_max' => 10,
        'armor_class' => 10,
    ]);
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_go-stdlib/play_privacy.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
)

// note is the response shape for a campaign note.
type note struct {
	NoteID     string `json:"note_id"`
	Text       string `json:"text"`
	Visibility string `json:"visibility"`
	Owner      string `json:"owner"`
}

// createNoteRequest binds the payload for creating a note.
type createNoteRequest struct {
	NoteID     string `json:"note_id"`
	Text       string `json:"text"`
	Visibility string `json:"visibility"`
}

// updateNoteRequest binds the payload for updating a note.
type updateNoteRequest struct {
	Text       string `json:"text"`
	Visibility string `json:"visibility"`
}

// noteListResponse is the shape returned by the note list endpoint.
type noteListResponse struct {
	Notes []note `json:"notes"`
}

// whisper is the response shape for a campaign whisper.
type whisper struct {
	WhisperID       string `json:"whisper_id"`
	FromCharacterID string `json:"from_character_id"`
	ToCharacterID   string `json:"to_character_id"`
	Text            string `json:"text"`
}

// createWhisperRequest binds the payload for creating a whisper.
type createWhisperRequest struct {
	WhisperID     string `json:"whisper_id"`
	ToCharacterID string `json:"to_character_id"`
	Text          string `json:"text"`
}

// whisperListResponse is the shape returned by the whisper list endpoint.
type whisperListResponse struct {
	Whispers []whisper `json:"whispers"`
}

// characterSheetResponse is the shape returned by the character sheet endpoint.
type characterSheetResponse struct {
	CharacterID      string `json:"character_id"`
	Owner            string `json:"owner"`
	Name             string `json:"name"`
	Class            string `json:"class"`
	Level            int    `json:"level"`
	ProficiencyBonus int    `json:"proficiency_bonus"`
	HPMax            int    `json:"hp_max"`
	ArmorClass       int    `json:"armor_class"`
}

// requireCampaignAccess authenticates the request and authorizes only the
// campaign owner or a campaign member. It returns the username, a bool that is
// true when the caller is the campaign DM, and a bool that is true on success.
// Unknown campaigns return 404.
func requireCampaignAccess(w http.ResponseWriter, r *http.Request, campaignID string) (string, bool, bool) {
	username, ok := bearerUsername(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return "", false, false
	}
	_, ok, err := loadUserByUsername(username)
	if err != nil {
		log.Printf("campaign access user query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return "", false, false
	}
	if !ok {
		writeError(w, http.StatusForbidden, "forbidden")
		return "", false, false
	}

	campaign, ok, err := queryPlayCampaign(campaignID)
	if err != nil {
		log.Printf("campaign access campaign query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return "", false, false
	}
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return "", false, false
	}
	if campaign.Owner == username {
		return username, true, true
	}

	out, err := dbQuery(fmt.Sprintf("SELECT 1 FROM play_campaign_members WHERE campaign_id=%s AND username=%s LIMIT 1;", sq(campaignID), sq(username)))
	if err != nil {
		log.Printf("campaign access member query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return "", false, false
	}
	var memberRows []struct {
		One int `json:"1"`
	}
	if err := json.Unmarshal(out, &memberRows); err != nil {
		log.Printf("campaign access member unmarshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return "", false, false
	}
	if len(memberRows) > 0 {
		return username, false, true
	}
	writeError(w, http.StatusForbidden, "forbidden")
	return "", false, false
}

// queryOwnedCharacterID returns the character_id owned by username in the
// campaign, if any. The caller must hold dbMu.
func queryOwnedCharacterID(campaignID, username string) (string, bool, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT character_id FROM play_campaign_members WHERE campaign_id=%s AND username=%s AND owner=%s LIMIT 1;", sq(campaignID), sq(username), sq(username)))
	if err != nil {
		return "", false, err
	}
	var rows []struct {
		CharacterID string `json:"character_id"`
	}
	if err := json.Unmarshal(out, &rows); err != nil {
		return "", false, err
	}
	if len(rows) == 0 {
		return "", false, nil
	}
	return rows[0].CharacterID, true, nil
}

// queryCampaignNote loads a single campaign note by campaign and note id.
// The caller must hold dbMu. It returns nil if the note does not exist.
func queryCampaignNote(campaignID, noteID string) (*note, bool, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT note_id, text, visibility, owner FROM campaign_notes WHERE campaign_id=%s AND note_id=%s LIMIT 1;", sq(campaignID), sq(noteID)))
	if err != nil {
		return nil, false, err
	}
	var rows []note
	if err := json.Unmarshal(out, &rows); err != nil {
		return nil, false, err
	}
	if len(rows) == 0 {
		return nil, false, nil
	}
	return &rows[0], true, nil
}

// queryCampaignWhisper loads a single campaign whisper by campaign and whisper id.
// The caller must hold dbMu. It returns nil if the whisper does not exist.
func queryCampaignWhisper(campaignID, whisperID string) (*whisper, bool, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT whisper_id, from_character_id, to_character_id, text FROM campaign_whispers WHERE campaign_id=%s AND whisper_id=%s LIMIT 1;", sq(campaignID), sq(whisperID)))
	if err != nil {
		return nil, false, err
	}
	var rows []whisper
	if err := json.Unmarshal(out, &rows); err != nil {
		return nil, false, err
	}
	if len(rows) == 0 {
		return nil, false, nil
	}
	return &rows[0], true, nil
}

// createNoteHandler lets authenticated campaign members create a note. The
// owner is derived from the authenticated actor.
func createNoteHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")

	username, _, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	var req createNoteRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.NoteID == "" || req.Text == "" || (req.Visibility != "private" && req.Visibility != "party") {
		writeError(w, http.StatusBadRequest, "invalid note")
		return
	}

	exists, err := queryExists(fmt.Sprintf("SELECT 1 FROM campaign_notes WHERE campaign_id=%s AND note_id=%s LIMIT 1;", sq(campaignID), sq(req.NoteID)))
	if err != nil {
		log.Printf("create note duplicate query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if exists {
		writeError(w, http.StatusConflict, "note already exists")
		return
	}

	if err := dbExec(fmt.Sprintf("INSERT INTO campaign_notes (campaign_id, note_id, text, visibility, owner, sort_order) VALUES (%s, %s, %s, %s, %s, COALESCE((SELECT MAX(sort_order) FROM campaign_notes WHERE campaign_id=%s), 0) + 1);",
		sq(campaignID), sq(req.NoteID), sq(req.Text), sq(req.Visibility), sq(username), sq(campaignID))); err != nil {
		log.Printf("create note insert error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusCreated, note{
		NoteID:     req.NoteID,
		Text:       req.Text,
		Visibility: req.Visibility,
		Owner:      username,
	})
}

// listNotesHandler returns campaign notes in creation order. The DM sees all
// notes; players see all party notes and only their own private notes.
func listNotesHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")

	username, isDM, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	out, err := dbQuery(fmt.Sprintf("SELECT note_id, text, visibility, owner FROM campaign_notes WHERE campaign_id=%s ORDER BY sort_order;", sq(campaignID)))
	if err != nil {
		log.Printf("list notes query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	var rows []note
	if err := json.Unmarshal(out, &rows); err != nil {
		log.Printf("list notes unmarshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	notes := make([]note, 0, len(rows))
	for _, n := range rows {
		if isDM || n.Visibility == "party" || n.Owner == username {
			notes = append(notes, n)
		}
	}
	if notes == nil {
		notes = []note{}
	}

	writeJSON(w, http.StatusOK, noteListResponse{Notes: notes})
}

// getNoteHandler returns a single note when readable. Private notes return 403
// to campaign members who are not the owner; the DM can read every note.
func getNoteHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")
	noteID := r.PathValue("note_id")

	username, isDM, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	n, ok, err := queryCampaignNote(campaignID, noteID)
	if err != nil {
		log.Printf("get note query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "note not found")
		return
	}

	if !isDM && n.Visibility == "private" && n.Owner != username {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	writeJSON(w, http.StatusOK, *n)
}

// updateNoteHandler lets only the note owner update text and visibility. The
// DM may read all notes but does not override ownership.
func updateNoteHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")
	noteID := r.PathValue("note_id")

	username, _, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	n, ok, err := queryCampaignNote(campaignID, noteID)
	if err != nil {
		log.Printf("update note query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "note not found")
		return
	}

	if n.Owner != username {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	var req updateNoteRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Text == "" || (req.Visibility != "private" && req.Visibility != "party") {
		writeError(w, http.StatusBadRequest, "invalid note")
		return
	}

	if err := dbExec(fmt.Sprintf("UPDATE campaign_notes SET text=%s, visibility=%s WHERE campaign_id=%s AND note_id=%s;",
		sq(req.Text), sq(req.Visibility), sq(campaignID), sq(noteID))); err != nil {
		log.Printf("update note update error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusOK, note{
		NoteID:     n.NoteID,
		Text:       req.Text,
		Visibility: req.Visibility,
		Owner:      n.Owner,
	})
}

// createWhisperHandler lets a campaign player with an owned character send a
// whisper to another campaign member's character.
func createWhisperHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")

	username, _, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	user, ok, err := loadUserByUsername(username)
	if err != nil {
		log.Printf("create whisper user query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok || user.Role != "player" {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	fromCharacterID, ok, err := queryOwnedCharacterID(campaignID, username)
	if err != nil {
		log.Printf("create whisper owned character query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	var req createWhisperRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.WhisperID == "" || req.ToCharacterID == "" || req.Text == "" {
		writeError(w, http.StatusBadRequest, "invalid whisper")
		return
	}

	toMemberExists, err := queryExists(fmt.Sprintf("SELECT 1 FROM play_campaign_members WHERE campaign_id=%s AND character_id=%s LIMIT 1;", sq(campaignID), sq(req.ToCharacterID)))
	if err != nil {
		log.Printf("create whisper recipient query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !toMemberExists {
		writeError(w, http.StatusBadRequest, "invalid whisper")
		return
	}

	exists, err := queryExists(fmt.Sprintf("SELECT 1 FROM campaign_whispers WHERE campaign_id=%s AND whisper_id=%s LIMIT 1;", sq(campaignID), sq(req.WhisperID)))
	if err != nil {
		log.Printf("create whisper duplicate query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if exists {
		writeError(w, http.StatusConflict, "whisper already exists")
		return
	}

	if err := dbExec(fmt.Sprintf("INSERT INTO campaign_whispers (campaign_id, whisper_id, from_character_id, to_character_id, text, sort_order) VALUES (%s, %s, %s, %s, %s, COALESCE((SELECT MAX(sort_order) FROM campaign_whispers WHERE campaign_id=%s), 0) + 1);",
		sq(campaignID), sq(req.WhisperID), sq(fromCharacterID), sq(req.ToCharacterID), sq(req.Text), sq(campaignID))); err != nil {
		log.Printf("create whisper insert error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusCreated, whisper{
		WhisperID:       req.WhisperID,
		FromCharacterID: fromCharacterID,
		ToCharacterID:   req.ToCharacterID,
		Text:            req.Text,
	})
}

// listWhispersHandler returns campaign whispers in creation order. The DM sees
// all whispers; players see only whispers involving their owned character.
func listWhispersHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")

	username, isDM, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	ownedCharacterID := ""
	if !isDM {
		id, ok, err := queryOwnedCharacterID(campaignID, username)
		if err != nil {
			log.Printf("list whispers owned character query error: %v", err)
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
		if ok {
			ownedCharacterID = id
		}
	}

	out, err := dbQuery(fmt.Sprintf("SELECT whisper_id, from_character_id, to_character_id, text FROM campaign_whispers WHERE campaign_id=%s ORDER BY sort_order;", sq(campaignID)))
	if err != nil {
		log.Printf("list whispers query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	var rows []whisper
	if err := json.Unmarshal(out, &rows); err != nil {
		log.Printf("list whispers unmarshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	whispers := make([]whisper, 0, len(rows))
	for _, w := range rows {
		if isDM || w.FromCharacterID == ownedCharacterID || w.ToCharacterID == ownedCharacterID {
			whispers = append(whispers, w)
		}
	}
	if whispers == nil {
		whispers = []whisper{}
	}

	writeJSON(w, http.StatusOK, whisperListResponse{Whispers: whispers})
}

// getCharacterSheetHandler returns the basic character sheet for a campaign
// character. Only the character owner and the campaign DM may read it.
func getCharacterSheetHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")
	characterID := r.PathValue("character_id")

	username, isDM, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	member, ok, err := queryPlayCampaignMember(campaignID, characterID)
	if err != nil {
		log.Printf("character sheet member query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "character not found")
		return
	}

	if !isDM && member.Owner != username {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	writeJSON(w, http.StatusOK, characterSheetResponse{
		CharacterID:      characterID,
		Owner:            member.Owner,
		Name:             member.Name,
		Class:            member.Class,
		Level:            1,
		ProficiencyBonus: 2,
		HPMax:            10,
		ArmorClass:       10,
	})
}

// createMessageRequest binds the payload for a new campaign chat message.
type createMessageRequest struct {
	Text string `json:"text"`
}

// createMessageResponse is the exact shape returned after creating a chat
// message. The kind is always "chat" and the actor is the authenticated user.
type createMessageResponse struct {
	Kind  string `json:"kind"`
	Actor string `json:"actor"`
	Text  string `json:"text"`
}

// createMessageHandler lets authenticated campaign members post a chat message.
// Spectator tokens are rejected because requireCampaignAccess only accepts
// session bearer tokens, returning 401 for any other bearer scheme.
func createMessageHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")

	username, _, ok := requireCampaignAccess(w, r, campaignID)
	if !ok {
		return
	}

	var req createMessageRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Text == "" {
		writeError(w, http.StatusBadRequest, "invalid message")
		return
	}

	if err := dbExec(fmt.Sprintf("INSERT INTO campaign_messages (campaign_id, actor, kind, text, sort_order) VALUES (%s, %s, 'chat', %s, COALESCE((SELECT MAX(sort_order) FROM campaign_messages WHERE campaign_id=%s), 0) + 1);",
		sq(campaignID), sq(username), sq(req.Text), sq(campaignID))); err != nil {
		log.Printf("create message insert error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusCreated, createMessageResponse{
		Kind:  "chat",
		Actor: username,
		Text:  req.Text,
	})
}
